Initial implementation

itsjeyd opened the pull request:
## VectorDraw XBlock

This XBlock allows course authors to define vector drawing exercises. To complete a vector drawing exercise, students must correctly place a set of predefined vectors (and points) on a board.

All data associated with this XBlock is stored in XBlock fields. Vectors, points, and expected result for an exercise must be defined using JSON, but we are also planning on adding WYSIWYG functionality to Studio that will allow authors to specify vectors and expected results by drawing them (cf. below). *This functionality will not replace existing functionality for specifying elements that are part of the exercise; it will instead be offered as an alternative method for creating content.* For a given exercise, student state consists of:

- the set of vectors and points that were present on the board when student last submitted an answer
- a list of checks to perform in order to obtain grade
- result returned by the grader for the most recent answer (which contains info about correctness of answer and feedback message).

The amount of data to store for an exercise depends on how complex it is (i.e., how many vectors and points there are in total); based on [examples](https://github.com/open-craft/jsinput-vectordraw/tree/master/Notes_and_Examples) from an existing JSInput implementation, it should not exceed 1-2K per problem, per student. Grading happens by comparing the set of elements on the board to a set of conditions specified by the course author ("expected result"). Minimally, this involves checking for each required element whether or not it is present in a given answer. Additional checks may include:

- checking coordinates of points
- checking coordinates of tail and/or tip of a vector
- checking the length of a vector
- checking the angle of a vector
- checking whether a line passes through one or more points.

None of these operations are very expensive computationally, and the grader short-circuits as soon as it encounters a failing check (there is currently no notion of a partially correct answer), so we don't expect there to be any performance bottlenecks. We haven't performed any load testing.

We'd like to get this XBlock on edx.org for Davidson College by mid-December.

**Testing**

1. Install the XBlock on Devstack via:

        pip install -r requirements.txt

2. Add `vectordraw` to the Advanced Module List in your course settings.
3. Add a "Vector Drawing" block to a unit.
4. Customize the Vector Drawing block. You can use one of the examples from the jsinput-vectordraw repository (e.g., [2_boxIncline_multiVector](https://raw.githubusercontent.com/open-craft/jsinput-vectordraw/master/Notes_and_Examples/2_boxIncline_multiVector/2_boxIncline.html)) for this purpose. Note that values of "Vectors", "Points", "Expected result" settings need to be valid JSON.

---

**Partner information** hosted on edx.org (Davidson College)
**Merge deadline** mid-December
**JIRA Story** [OSPR-935](https://openedx.atlassian.net/browse/OSPR-935)
**Confluence / Product Asset** N/A
**Sandbox URL** [LMS](http://pr10515.sandbox.opencraft.com/) ([Example](http://pr10515.sandbox.opencraft.com/courses/course-v1:OpenCraft+VD101+2015/courseware/be8fac39945d427c987af0bdbec200c3/ecb2b2ba39214ee1b272dd83e8ad2890/)), [Studio](http://studio.pr10515.sandbox.opencraft.com/) ([Example](http://studio.pr10515.sandbox.opencraft.com/container/block-v1:OpenCraft+VD101+2015+type@vertical+block@2ecf812afd434f359480c7852ce12fa0); see author notes below about a known glitch)
**Dependencies** N/A

---

**PR Author(s) Notes / To-Do**
* [x] Integration tests
* [x] Accessibility
* [x] Moving controls for interacting with the board above the active board (as per [this request](https://github.com/open-craft/xblock-vectordraw/pull/1#issuecomment-153837342)).
* [x] There's a glitch in Studio that keeps vectors from being rendered properly when first loading the page (you have to click "Reset" to get a proper preview). We'll have another opportunity to address this in the context of the next item.
* [x] WYSIWYG functionality for defining vectors and expected results

---

**Reviewers**
- [x] Code: @smarnach
- [x] Code: @nedbat
- [x] Code: @singingwolfboy
- [ ] Accessibility: @cptvitamin

smarnach commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,229 @@
+                return {'ok': False, 'msg': result}
+        return {'ok': True, 'msg': self.success_message}
+
+    def cfn(self, e, ans):
```
I don't think this function is needed for the XBlock version – it looks quite JSInput-specific.

itsjeyd replied:
@smarnach Ah, yes - missed that somehow when I was cleaning up. Thanks!

smarnach commented on vectordraw/static/js/src/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+        '      </div>',
+        '    <% } %>',
+        '</div>'
+    ].join('\n'));
```
I think we should render this template on the server side now that we can.  It only uses Studio settings and no user state, so it seems like a good fit for being included in the Django template.

smarnach commented on vectordraw/static/js/src/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+
+    // Initialization logic
+
+    $(function ($) {
```
I know this was inserted by xblock-sdk, but I don't quite get the point.  The XBlock init function is called after the DOM is ready anyway, so you might just as well insert the code directly.

smarnach commented on vectordraw/static/js/src/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+        checkXHR = $.post(checkHandlerUrl, JSON.stringify(state))
+            .success(function(response) {
+                updateStatus(response);
+            });
```
This is the same as `.success(updateStatus);`.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+    )
+
+    height = Integer(
+        display_name="",
```
Display name is missing.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+    )
+
+    bounding_box_size = Integer(
+        display_name="",
```
Here as well.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+
+    # User state
+    answer = Dict(scope=Scope.user_state)
+    result = Dict(scope=Scope.user_state)
```
A comment describing the content of these two fields would be nice.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+    def expected_result_json(self):
+        return json.loads(self.expected_result)
+
+    def resource_string(self, path):
```
You probably don't need this function when using the `ResourceLoader` class from xblock-utils.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+        fragment = Fragment()
+        fragment.add_content(loader.render_template('static/html/vectordraw.html', context))
+        fragment.add_css_url("//cdnjs.cloudflare.com/ajax/libs/font-awesome/4.3.0/css/font-awesome.min.css")
+        fragment.add_css(self.resource_string('static/css/vectordraw.css'))
```
You should use `loader.load_unicode()` here for consistency.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+        fragment.add_css_url("//cdnjs.cloudflare.com/ajax/libs/font-awesome/4.3.0/css/font-awesome.min.css")
+        fragment.add_css(self.resource_string('static/css/vectordraw.css'))
+        fragment.add_javascript_url("//cdnjs.cloudflare.com/ajax/libs/jsxgraph/0.98/jsxgraphcore.js")
+        fragment.add_javascript(self.resource_string("static/js/src/vectordraw.js"))
```
Ditto.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+        self.answer = dict(
+            vectors=data["vectors"],
+            points=data["points"]
+        )
```
You should validate the format of `data` before saving the answer.  This isjust HTTP request, so people could send you anything.  On the server, never trust anything that comes from the browser!

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+        self.result = result
+        # Publish grade data
+        score = 1 if result["ok"] else 0
+        self.runtime.publish(self, 'grade', dict(value=score, max_value=1))
```
Shouldn't you have `weight` somewhere here?  I'm not sure about this, maybe `weight` is considered automatically.

itsjeyd replied:
@smarnach Good question, I had to look this up myself. Based on info from the [edX Grading Guide for Developers](https://openedx.atlassian.net/wiki/display/TNL/edX+Grading+Guide+for+Developers), problem weights are applied when calculating grades, so we don't need to factor `weight` in at this level.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,312 @@
+                <vectordraw/>
+                </vertical_demo>
+             """),
+        ]
```
This scenarios doesn't make too much sense yet (though it would likely work, showing three copies of a vectordraw XBlock with all settings at their default values).

smarnach commented:
@itsjeyd Looks quite good so far!  I did a first pass through the code, and added a few comments.

So far I didn't do any manual testing – will do this in the next iteration.

itsjeyd commented:
@smarnach Thanks for reviewing! I addressed all of your comments, so this is ready for you to have another look.

itsjeyd commented:
@dseaton We have ported almost all of the features of jsinput-vectordraw to this XBlock. The only thing that is currently missing is support for custom checks. How important are these for your current/upcoming use cases? The reason I bring this up is that there is currently no good/standard solution for XBlocks to execute Python code entered by the user. In principle, it would be possible to add support for this by accepting custom Python code and running it in a sandbox (just like JSInput does), but this would likely take a significant chunk of time to implement, and might end up causing this project to exceed the time currently allocated for it. Let me know if it would make sense to defer adding support for custom checks to a future v2 of the Vector Drawing XBlock.

CC @antoviaque

smarnach commented on .travis.yml:
```diff
@@ -0,0 +1,19 @@
+  - 2.7
+before_install:
+  - export DISPLAY=:99
+  - sh -e /etc/init.d/xvfb start
```
This isn't really used yet, since there aren't any integration tests.  We can leave this here, since we will have to add integration tests anyway.  They will be a bit tricky to implement, but it won't be too bad.

smarnach commented on run_tests.py:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
```
This file isn't necessary for the unit tests as well, but good to have for the integration tests.

smarnach commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,345 @@
+        for check in answer['checks']:
+            check_data['check'] = check
+            check_fn = self.check_registry[check['check']]
+            args = [check_data[arg] for arg in inspect.getargspec(check_fn).args]
```
I just noticed this hack I introduced when adding points as a problem type.  It's main purpose was to stay backwards-compatible with existing external graders.  We might now be able to simply pass `check, vectors, points` as arguments to all graders.  (Up to you whether you want to change this.  it would require editing the parameter lists of all functions in this file, and updaing all the tests.)

nedbat replied:
btw, another way to do this would be to declare all checks with a `**unused` argument at the end, and then invoke it with `check_fn(**check_data)`.

smarnach commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,345 @@
+    Return error message for `check` targeting a vector from `vectors`.
+
+    If `check` does not define a custom error message, fall back on `default_message`.
+    """
```
Nice to have docstrings everywhere now!

nedbat replied:
It's pretty clear there are coordinated data structures in wide use throughout this module.  If there were a description of those structures at the top of the module, or a pointer to a description, it would make it easier to review, understand, and maintain this code.

smarnach commented on vectordraw/static/js/src/vectordraw.js:
```diff
@@ -0,0 +1,607 @@
+                if (noOptionSelected) {
+                    option.prop('selected', true);
+                    noOptionSelected = false;
+                }
```
I didn't realize you would need to do this now the template code is moved to the static template.  Instead of integrating this in the loop, you could also simply enable all options using

    this.element.find('.menu option').prop('disabled', false);

and then select the first one (if necessary) with

    this.element.find('.menu option:first-child').prop('selected', true);

smarnach commented on vectordraw/templates/xml/box_incline.xml:
```diff
@@ -0,0 +1,13 @@
+              bounding_box_size="8"
+              background_url="https://github.com/open-craft/jsinput-vectordraw/raw/master/Notes_and_Examples/2_boxIncline_multiVector/box_on_incline.png"
+              background_width="20"
+              vectors="[{&quot;name&quot;:&quot;N&quot;,&quot;description&quot;:&quot;Normal force - N&quot;,&quot;tail&quot;:[2,2],&quot;length&quot;:4,&quot;angle&quot;:45,&quot;render&quot;:true},{&quot;name&quot;:&quot;f&quot;,&quot;description&quot;:&quot;Friction - f&quot;,&quot;coords&quot;:[[-2,-5],[-1,-3]],&quot;render&quot;:false},{&quot;name&quot;:&quot;g&quot;,&quot;description&quot;: ...
```
This is unfortunately rather unreadable.  There is a way to turn attributes into XML tags, which might improve readability here (see [the `xml_node` parameter](http://edx.readthedocs.org/projects/xblock/en/latest/fields.html#xblock.fields.Field)).  Since we might change the fields and their format later on, it's probably not worthwhile changing this now, just something to keep in mind.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,387 @@
+        scope=Scope.content
+    )
+
+    vectors = String(
```
Another thing to keep in mind:  If we decide to keep this field in its current form, we eventually want to turn this field into one of type `List` to make the JSON serialisation implicit and to model better what it is.  We can't do this now, since `List` fields aren't handled by the `StudioEditableXBlockMixin`.

Same comment hold for the `points` field.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,387 @@
+            # template needs this info to determine whether it should add option
+            # for selecting point to dropdown menu:
+            if 'fixed' not in point:
+                point['fixed'] = True
```
Since you had to move part of the default value code on the server side, it makes sense to move all of it.  We generally want to have as much of the code in Python as possible, since that code is easier to test and maintain (and also because Python is not quite as insane a language as JS).

If there are any blockers to moving the default value code here, use your own judgement to decide whether it makes sense, but if it is straight-forward, I think it woulld be an improvement.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,387 @@
+        The primary view of the VectorDrawXBlock, shown to students
+        when viewing courses.
+        """
+        context['self'] = self
```
`context` might be `None` (see the parameter defintion).  You can simply add

    context = context or {}

before this line.

smarnach commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,387 @@
+        fragment.add_javascript_url(
+            "//cdnjs.cloudflare.com/ajax/libs/jsxgraph/0.98/jsxgraphcore.js"
+        )
+        fragment.add_javascript(loader.load_unicode("static/js/src/vectordraw.js"))
```
This file is 22k, so it's a bit too big for inlining.  Add it by URL instead (which requires moving it to the `public` directory).

itsjeyd replied:
@smarnach Just FYI, I ended up getting rid of the `static` directory entirely -- it seemed weird to have JS and CSS files spread out over two directories. So both CSS and JS files for this XBlock now live in `public/{css,js}`, and templates are stored in `templates/html`. This is mimics the directory layout of, e.g., problem-builder and xblock-drag-and-drop-v2.

smarnach replied:
@itsjeyd The three directories are for different purposes.  `templates` is for templates of any kind, `static` is for content that is used as-is by the Python code, but not served directly by the runtime and `public` is for static content that is served directly from the filesystem by the runtime.

If you have a small JS snippet you want to inline, you would put it in `static`, since it's only loaded by the Python code in your XBlock.  If you want to include the JS code by URL, it needs to be served directly from the filesystem, so it has to go into `public`.

smarnach commented:
We will need more tests for this.  Part of the motivation for moving from JSInput to XBlock format is that XBlocks can be tested more easily.  We generally add unit tests for everything that can be unit-tested, and in addition integration tests to test overall functionality and cover parts that aren't reasonably unit-testable.  Integration tests will be a bit tricky to implement for vector drawing, but it shouldn't be too bad.

smarnach commented on vectordraw/static/js/src/vectordraw.js:
```diff
@@ -0,0 +1,607 @@
+        });
+
+        function getImageRatio(bg, callback) {
+            $('<img/>').attr('src', bg.src).load(function(){
```
I think this code might be causing the issue in Studio with the widget not being correctly initialised on first load.  The load event is kind of tricky and incosistent – see e.g. the caveats in http://api.jquery.com/load-event/.

smarnach commented:
The styling of the buttons in the new version is still a bit weird.  The issue is that the JSInput version runs inside an iframe, so it is shielded from the LMS styles, while the XBlock version inherits all styles from the LMS.  This results in the buttons having a weird mixture of style from the old version and the LMS.

The new version doesn't have to look identical to the old version, but it should at least look like it was design that way intentionally rather than by accident. :)  I also liked that the undo and redo buttons were next to each other in the old version.

smarnach commented:
Even with all settings being identical, the scaling of the scaling and aspect ratio of the board is different in the new version.  (I don't know why that's the case either.)

dseaton commented:
@itsjeyd Custom checks played a small role in our initial release. We can likely table it for now, and use the JSinput version if necessary to have them. The main use that I remember was for comparing the length of two vectors. For example, two vectors might be required to have the same length. Another might be that a resultant vector have properties of two other vectors.

Someone please point me to an example when ready to test.

smarnach commented:
@itsjeyd I've looked through the code in a bit more detail and tested manually.  It's working really well already, so well done!  I'm sorry I still had to give you quite a bit of work. :/

smarnach commented:
@dseaton It would likely be possible (and siginificantly less work) to add some of these custom tests as canned tests instead.

antoviaque commented:
@itsjeyd Btw to allow @dseaton to test, it might be worth setting up a sandbox with the XBlock on it. Since we also need to get the edX upstream review started, it would be worth opening a PR against edx/edx-platform with an additional requirement pointing to this repo (use the branch rather than commit # for now, to be able to update that sandbox easily).

itsjeyd commented:
@smarnach Thanks for the feedback!

> The styling of the buttons in the new version is still a bit weird. I also liked that the undo and redo buttons were next to each other in the old version.

I adjusted the styling a little. Undo and redo buttons are next to each other now, and I'm using a much lighter gradient on hover:

![button-styles-adjusted](https://cloud.githubusercontent.com/assets/961441/10934868/ae115936-82e2-11e5-8e99-0d195e486ae6.png)

> Even with all settings being identical, the scaling of the scaling and aspect ratio of the board is different in the new version. (I don't know why that's the case either.)

I'm not sure what's causing this either. Are there any circumstances under which this would be a problem?

---

As for the Studio glitch, I looked into it more and I'm not sure if the `load` event is (the only thing) to blame in this case. I had a look at the HTML of the `.jxgboard` node *before* clicking "Reset"; turns out that the `svg` element that holds the contents of the board has both `width` and `height` set to `0`. (After clicking "Reset", these attributes are correctly set to expected values.) Also, vectors and points that are not initially hidden are missing attributes that are relevent for drawing them (such as `x1`, `y1`, `x2`, `y2` for vectors and `cx`, `cy`, `rx`, `ry` for points/ellipses). It's almost as if the initialization code for the board gets run "too early" in Studio (i.e., before all of the information that's relevant for drawing the board correctly is available).

dseaton commented:
@antoviaque @smarnach @itsjeyd Xavier and I discussed possibly changing the interface such that buttons and displays appeared above the active board. This would allow us to avoid overflow issues (that the technical term?) when dealing with horizontal spacing in edX - which changes from time to time. :)

antoviaque commented:
@dseaton Thanks for mentioning this now - the current scope @itsjeyd is working with only includes the conversion of the existing tool "as is", but it would definitely make sense to do the layout rework at this point. I'll include a story about this for an upcoming sprint.

itsjeyd commented:
@dseaton We now have a sandbox for this that you can use to test. Please see the updated description of this PR for Studio and LMS links.

sarina commented:
Also it doesn't work when I get the right answer - the check button does nothing and still displays the answer from the previous "check".

![screen shot 2015-11-05 at 11 51 47 am](https://cloud.githubusercontent.com/assets/1985317/10975190/ae738864-83b3-11e5-998d-3908996975c0.png)

sarina commented:
I'm really concerned about accessibility, but also student feedback. When I hit "Check", it only tells me if one of my vectors is wrong, no feedback about the other two. There's no nudging me to the right answer. There's no show answer to show me the right answer.

![screen shot 2015-11-05 at 11 47 59 am](https://cloud.githubusercontent.com/assets/1985317/10975067/19ee2f6e-83b3-11e5-833a-b5a19897ce22.png)

Is "N" right? Is "g" right? I don't know what "f" is, what do I do?

itsjeyd commented:
@sarina Thanks for the feedback.

> Also it doesn't work when I get the right answer - the check button does nothing and still displays the answer from the previous "check".

This should be fixed now. I'm reprovisioning the sandbox, so it won't be available for a while.

itsjeyd commented:
@sarina PR description updated to address the questions you sent via email. I'm reproducing the info here for convenience:

> All data associated with this XBlock is stored in XBlock fields. Vectors, points, and expected result for an exercise must be defined using JSON, but we are also planning on adding WYSIWYG functionality to Studio that will allow users to specify vectors and expected results by drawing them (cf. below). *This functionality will not replace existing functionality for specifying elements that are part of the exercise; it will instead be offered as an alternative method for creating content.* For a given exercise, student state consists of:

> - the set of vectors and points that were present on the board when student last submitted an answer
> - a list of checks to perform in order to obtain grade
> - result returned by the grader for the most recent answer (which contains info about correctness of answer and feedback message).

> The amount of data to store for an exercise depends on how complex it is (i.e., how many vectors and points there are in total); based on [examples](https://github.com/open-craft/jsinput-vectordraw/tree/master/Notes_and_Examples) from an existing JSInput implementation, it should not exceed 1-2K per problem, per student. Grading happens by comparing the set of elements on the board to a set of conditions specified by the course author ("expected result"). Minimally, this involves checking for each required element whether or not it is present in a given answer. Additional checks may include:

> - checking coordinates of points
> - checking coordinates of tail and/or tip of a vector
> - checking the length of a vector
> - checking the angle of a vector
> - checking whether a line passes through one or more points.

> None of these operations are very expensive computationally, and the grader short-circuits as soon as it encounters a failing check (there is currently no notion of a partially correct answer), so we don't expect there to be any performance bottlenecks. We haven't performed any load testing.

Let me know if there's anything else you need to know. Will the review for this happen at the same time as the review for the Active Table XBlock?

And thanks again for taking the time last week to talk about a11y with us!

itsjeyd commented:
@sarina

> I'm really concerned about [...] student feedback. When I hit "Check", it only tells me if one of my vectors is wrong, no feedback about the other two. There's no nudging me to the right answer.

Error messages for failed checks are customizable; what you saw on the sandbox was a default message. It is up to course authors how much information they want to provide/reveal when a specific check fails. Also, as mentioned in the previous comment, the grader stops at the first failing check. It returns the error message that is associated with that specific check, and that is what we currently display in the UI. Once the student successfully fixes the issue, they are either completely done (because all of the remaining checks pass), or they will get feedback about another issue that affects their answer when hitting "Check". This allows them to fix their answer step by step, until they get to the correct result.

smarnach commented:
@itsjeyd Looks nice!  I like that the integration tests are really comprehensive – this will be really helpful for the further development of the XBlock.  I added a couple of stylistic comments for the tests – feel free to ignore them at will.  :+1: after addressing the comments, provided the tests are still passing.

itsjeyd commented:
@smarnach Thanks for the comments! I addressed them all :)

itsjeyd commented:
@sarina @cptvitamin @dseaton FYI, our [sandbox](http://pr10515.sandbox.opencraft.com/) for this is up and running again (Studio link [here](http://studio.pr10515.sandbox.opencraft.com/)). It currently has five examples for vector drawing problems.

CC @antoviaque

dseaton commented:
@itsjeyd @antoviaque How do I access the sandbox? i.e., credentials.

cptvitamin commented:
@itsjeyd Thanks for getting that sandbox up.  I left some comments on the JIRA ticket: https://openedx.atlassian.net/browse/AC-241

Let me know if you have any questions.  Would be glad to help.

itsjeyd commented:
@dseaton You can use the standard accounts (`staff`, `audit`, etc.) and associated password for logging in to the sandbox.

antoviaque commented:
@dseaton More specificaly, the password for these default accounts is 'edx', and the email `<account-name>@example.com` (for example, staff@example.com). Don't hesitate if you have any issue login in.

dseaton commented:
@itsjeyd @antoviaque Just accessed the examples. If desired, I can give you some others to convert that highlight other features.

Note, in Safari, each of the problems has trouble loading initially. The canvas is non responsive, and no image loads. If you hit reset, things become responsive and the image appears. But the click and drag for creating vectors seems to still have some trouble.

antoviaque commented:
@dseaton Thanks for the testing, and the bug report : )

For extra examples, that could be a good thing yes - that might help doing further testing and find out other issues, before this gets published. I'll add to our backlog to test them (or feel free to set them up directly on the sandbox, actually - just let us know so we make sure we save them).

dseaton commented:
@itsjeyd @antoviaque Not sure how I would code these, so for now, I am going to put key examples that should make it into the tests. Offline, tell me and I can give someone studio access.

Rotational Motion:
https://courses.edx.org/courses/course-v1:DavidsonNext+Phy_APccx+2T2015/courseware/Rotational_Motion/LC1_Defining_Rotational_Motion/10

Atomic Transitions:
https://courses.edx.org/courses/course-v1:DavidsonNext+Phy_APccx+2T2015/courseware/Atomic_Transitions/LC2_Calculate_Energy_Required_for_Transitions/3


Making Measurements:
Standard:  https://courses.edx.org/courses/course-v1:DavidsonNext+Phy_APccx+2T2015/courseware/Diffraction_and_Interference/LC2_Two_Source_Interference/8

Laboratory - measuring diffraction patterns
https://courses.edx.org/courses/course-v1:DavidsonNext+Phy_APccx+2T2015/courseware/Diffraction_and_Interference/LC4_Single_Slit_Interference/3

antoviaque commented:
@dseaton Thanks! For the access to Studio, the right person would probably be @itsjeyd . Tim, is your email on edx.org your OpenCraft one?

itsjeyd commented:
@antoviaque Yes, it is.

antoviaque commented:
@itsjeyd Thanks! @dseaton so this is tim AT opencraft.com (our emails are all firstname@opencraft.com)

dseaton commented:
@itsjeyd Studio access granted on edge. Note, the courses are identical on edge and edx. 

Let me know if you can't find anything.

antoviaque commented:
@dseaton Btw, for the sample exercises, given that we have largely exceeded the time quota on this task, we don't really have time left to allocate to recreating them with the XBlock. Is it ok if we skip this, or maybe include it in follow-up work?

itsjeyd commented:
Hi @sarina, could you give me an update on when the review for this XBlock is going to happen? Since the goal is to have it merged by Dec 15th, we'd need to get feedback early this week to have a chance to match that deadline. Thanks!

dseaton commented:
@antoviaque Fine to skip for now, but I do think we want to get those examples into a test course. I can likely do this next week. Julie and I owe you an email.

sarina commented:
I've been waiting for @nedbat to be free for a review but he is continuously busy. Ned, will you be able to review this XBlock this week? Do I need to pull in Joel or someone else?

antoviaque commented:
@dseaton Sounds good, thanks for understanding! And since this is important, we can still try to accomodate it to ensure it still gets proper testing - if you try to enter the exercises yourself that definitely help, but we could still assist; if you have blockers or questions let me know.

nedbat commented:
I'll be looking at this tomorrow.  I'm impressed by the thoroughness of the review it has already gotten!

nedbat commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+django>=1.4, <1.5
```
This seems like it should be in test-requirements.txt

nedbat commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,345 @@
+        return _errmsg('Vector {name} does not start at correct point.', check, vectors)
+
+
+def check_tip(check, vectors):
```
This function has a great similarity to check_tail.  Is there a way to share code between the two?

nedbat commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,345 @@
+    """
+    line = vectors[check['vector']]
+    tolerance = check.get('tolerance', 1.0)
+    points = check.get('expected')
```
Seems like 'expected' is a required key, so use `check['expected']`

nedbat commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,345 @@
+    """
+    point = points[check['point']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check.get('expected')
```
Seems like 'expected' is a required key, so use check['expected']

nedbat commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,541 @@
+            "Note that you can also use the WYSIWYG editor below to opt in and out of checks "
+            "for individual vectors. "
+            "If you use the WYSIWYG editor at all, any changes you make here "
+            "will be overwritten when saving."
```
This help string doesn't provide guidance about the syntax of the string.

nedbat commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,541 @@
+            default_vector_style = default_vector['style']
+            if 'style' in vector:
+                default_vector_style.update(vector['style'])
+                del vector['style']
```
For these cases where you have to use and remove a value from a dict, `d.pop(k)` is useful.

nedbat commented on vectordraw/templates/html/vectordraw.html:
```diff
@@ -0,0 +1,130 @@
+
+  {% if self.description %}
+    <div class="vectordraw-description">
+      {{ self.description|safe }}
```
Why is the description considered HTML?  Shouldn't it be escaped like other things?

itsjeyd replied:
@nedbat We allow authors to format the description using a rich editor (TinyMCE), so it may contain HTML tags. (This is done by [initalizing the field with `multiline_editor="html"`](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/vectordraw.py#L45) and [attaching a TinyMCE instance to the field](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/public/js/studio_edit.js#L44) on the client.)

nedbat commented on vectordraw/templates/html/vectordraw_edit.html:
```diff
@@ -0,0 +1,372 @@
+                        aria-describedby="{{field.name}}-help"
+                        rows=10 cols=70>{{field.value}}</textarea>
+            {% else %}
+              Unsupported field type. This setting cannot be edited.
```
Everything else is nicely marked with `{% trans %}`, you might as well get this one too.

nedbat commented:
No major concerns.  I haven't read the JavaScript code (not my skill-set).  :+1:

itsjeyd commented:
@nedbat Thanks for reviewing! I addressed your comments, the changes are here: d2a9f6e23dbcc116b4b7d5ceb51874850075fc95.

And thank you @sarina for scheduling the review!

nedbat commented:
Will someone be reviewing the JavaScript code?

itsjeyd commented:
@nedbat Is this a question for me/OpenCraft or for @sarina? :)

nedbat commented:
@singingwolfboy will be doing the JavaScript review.

itsjeyd commented:
@nedbat Great, thanks for the heads-up!

singingwolfboy commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+pyyaml
+-e git+https://github.com/edx/XBlock.git@tag-master-2015-05-22#egg=XBlock
+-e git+https://github.com/edx/xblock-utils.git@b4f9b51146c7fafa12f41d54af752b8f1516dffd#egg=xblock-utils
```
Why the `-e` flags? I'm pretty sure that we don't want these in requirements files. (edx-platform has a lot of these `-e` dependencies, and they've caused us a fair amount of trouble in the past -- we're trying to get away from that.)

singingwolfboy commented on setup.py:
```diff
@@ -0,0 +1,40 @@
+    ],
+    install_requires=[
+        'XBlock',
+        'xblock-utils',
```
Looks like `pyyaml` is missing from this list. It's listed in the `requirements.txt` file, and not here. (Maybe you should unify these somehow?)

itsjeyd replied:
@singingwolfboy Thanks for catching that. The Vector Drawing XBlock itself actually doesn't make use of `pyyaml` (it only depends on it because it extends `XBlock`). So instead of adding it here I removed `pyyaml` from `requirements.txt` and bumped the `XBlock` version to `0.4.2` (which lists `pyyaml` in its [`requirements.txt` file](https://github.com/edx/XBlock/blob/bfa7d968157c188d8e4d80e5ded839248c490ce1/requirements.txt#L2)).

singingwolfboy commented on run_tests.py:
```diff
@@ -0,0 +1,52 @@
+    os.environ.setdefault('DJANGO_LIVE_TEST_SERVER_ADDRESS', 'localhost:8081-8099')
+
+    try:
+        os.mkdir('var')
```
Why are you making this directory? What is it used for?

itsjeyd replied:
@singingwolfboy The workbench uses the `var` directory [for database and log files](https://github.com/edx/xblock-sdk/commit/f8bff8ba9279ae5bcbd99eed0731f39273a237ae). Creating it here to ensure that the tests don't fail because it's missing.

smarnach replied:
I think we should fix this in the workbench eventually.  It's kind of stupid that workbench simply errors out if the directory doesn't exist, so all XBlocks have to create it.

nedbat replied:
Please write issues for things like this.  I've done this one: https://github.com/edx/xblock-sdk/issues/99

singingwolfboy commented on test-requirements.txt:
```diff
@@ -0,0 +1,5 @@
+django>=1.4, <1.5
```
edx-platform is now on Django 1.8. Is that going to be a problem?

singingwolfboy commented on test-requirements.txt:
```diff
@@ -0,0 +1,5 @@
+django>=1.4, <1.5
+-r requirements.txt
+-e git+https://github.com/edx/xblock-sdk.git@629650ed6fc1b77d1a5287debeace0b29db7c0fd#egg=xblock-sdk
```
Remove the `-e` unless you have a reason to keep it.

itsjeyd replied:
@singingwolfboy Without the `-e` flag it doesn't seem possible to install test requirements for xblock-sdk (`pip install -r $VIRTUAL_ENV/src/xblock-sdk/requirements/base.txt` and `pip install -r $VIRTUAL_ENV/src/xblock-sdk/requirements/test.txt` in `travis.yml` fail with a "No such file or directory" error), so I'm keeping it for now.

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+        else:
+            self.fail(errmsg)
+
+    def check_hidden_text(self, selector, expected_text):
```
Assertion methods generally involve the word `assert` in the method name.

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+            src = background.get_attribute("xlink:href")
+            self.assertEquals(src, "https://github.com/open-craft/jsinput-vectordraw/raw/master/Notes_and_Examples/2_boxIncline_multiVector/box_on_incline.png")
+            alt = background.get_attribute("alt")
+            self.assertEquals(alt, "A very informative description")
```
This description seems like it should be configurable.

itsjeyd replied:
@singingwolfboy The description is configurable in Studio via the `background_description` field (cf. [`editable_fields`](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/vectordraw.py#L234) and [`drawBackground`](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/public/js/vectordraw.js#L55)); this is just a [dummy value](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/tests/integration/xml/custom.xml#L15).

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+            "points": json.dumps([]),
+            "expected_result": json.dumps({})
+        },
+    )
```
I'm not sure how I feel about putting so much information into the `@data` decorator. Can you just make these separate tests?

itsjeyd replied:
@singingwolfboy I'm not sure we'd gain much by splitting this test. IIUC, I'd be creating two separate tests that would define the data to test on and then call another method that would contain the actual testing logic. Basically, I'd be replicating the functionality that I get for free from the `@data` decorator.

singingwolfboy commented:
Looks to me like all of the function's in the Grader's `check_registry` return a string when the check fails, and None when the check succeeds. This feels very unintuitive to me -- it means that the caller must always check the return value when calling a check function to see if it succeeded or fails. Why not have the check function raise a `ValueError` if there is a failure, or even an `AssertionError`? That way, the caller doesn't have to check return values, but can instead assume that everything succeeds by default. If the caller wants to handle the case where the check function fails, it can simply call it within a `try`/`except` block, and handle the exception that way.

singingwolfboy commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,381 @@
+        return Vector(self.name, self.tip.x, self.tip.y, self.tail.x, self.tail.y)
+
+
+class Grader(object):
```
Looks to me like this `Grader` class exists for one reason only: its `grade()` method. Why define the class at all? Why not just pull the `grade()` method out into a standalone function, and call that?

itsjeyd replied:
@singingwolfboy You're right about the fact that the `grade` method is the only thing being called from outside. But I think it makes sense to encapsulate the grading logic in a class, because otherwise we'd end up with either one long function that has all of the logic from the `Grader` class, or multiple functions that you wouldn't know are related just by looking at them (i.e., you'd have to read the code in more detail to understand the relationship between the functions).

singingwolfboy commented on vectordraw/grader.py:
```diff
@@ -0,0 +1,381 @@
+            result = check_fn(*args)
+            if result:
+                return {'ok': False, 'msg': result}
+        return {'ok': True, 'msg': self.success_message}
```
This return type seems wrong to me. Python allows functions to return multiple values. If you want the function to report both whether the student's answer was correct or not, *and* a message to report to the student, you can just return those two values, rather than combining them into a dictionary. I also don't like the word `ok` in this context: it's ambiguous whether you mean "this code ran successfully vs. failed" or if you mean "the student's response is marked as passing vs failing".

itsjeyd replied:
@singingwolfboy Definitely agree on `ok` not being the best choice in this context; I changed it to `correct` to make it clear that it refers to the student's answer being marked as correct vs. incorrect. In terms of the return type I think it makes sense for the grader to return the result in a format that we can use *as is* to both persist information about the most recent result obtained by the student (in the [JSON handler](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/vectordraw.py#L530)), and to display appropriate feedback to the student in the LMS (cf. [`updateStatus`](https://github.com/open-craft/xblock-vectordraw/blob/initial-implementation/vectordraw/public/js/vectordraw.js#L592)). (In other words, we'd be constructing that dict anyway if the grader returned separate values.)

singingwolfboy commented:
Based on the Python code, it looks like you're constructing your own representations of mathematical points and vectors, and implementing code to do operations on them (like finding distances). Why not use [numpy](http://www.numpy.org/) instead? You can use [`numpy.array`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html) to represent points and vectors, and there's a lot of functionality that numpy will give you for free with that. It means less of your own code, and fewer tests as well.

However, I just noticed that the merge deadline for this pull request is rather tight, so I doubt you'll be able to take this suggestion right now. A suggestion for the future.

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+
+    var fields = [];
+    var tinyMceAvailable = (typeof $.fn.tinymce !== 'undefined'); // Studio includes a copy of tinyMCE and its jQuery plugin
+    var errorMessage = gettext("This may be happening because of an error with our server or your internet connection. Try refreshing the page or making sure you are online.");
```
Rather than "Try refreshing the page or making sure you are online", I would suggest "Make sure you are online, and try refreshing the page."

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+            val: function() {
+                var val = $field.val();
+                // Cast values to the appropriate type so that we send nice clean JSON over the wire:
+                if (type == 'boolean')
```
Never use `==` in Javascript, always use `===`

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+            url: handlerUrl,
+            data: JSON.stringify(data),
+            dataType: "json",
+            global: false  // Disable Studio's error handling that conflicts with studio's notify('save') and notify('cancel') :-/
```
If you just want to disable Studio's error handling, you can set `notifyOnError: false`. See https://github.com/edx/edx-platform/blob/master/cms/static/coffee/src/main.coffee#L22

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+            if (jqXHR.responseText) { // Is there a more specific error message we can show?
+                try {
+                    errorMessage = JSON.parse(jqXHR.responseText).error;
+                    if (typeof errorMessage === "object" && errorMessage.messages) {
```
You could use [`_.isObject()`](http://underscorejs.org/#isObject) if you want -- it's shorter and clearer.

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+                    errorMessage = JSON.parse(jqXHR.responseText).error;
+                    if (typeof errorMessage === "object" && errorMessage.messages) {
+                        // e.g. {"error": {"messages": [{"text": "Unknown user 'bob'!", "type": "error"}, ...]}} etc.
+                        errorMessage = $.map(errorMessage.messages, function(msg) { return msg.text; }).join(", ");
```
You could use [`_.pluck()`](http://underscorejs.org/#pluck) if you want -- it's shorter and clearer. In this case, you would do:
```js
var errorMessages = _.pluck(errorMessage.messages, "text");
errorMessage = errorMessages.join(", ");
```

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+        save: function(data) {
+            var values = {};
+            var notSet = []; // List of field names that should be set to default values
+            for (var i in fields) {
```
Why not use [`_.each()`](http://underscorejs.org/#each) or [`$.each()`](http://api.jquery.com/jQuery.each/) instead?

singingwolfboy commented on vectordraw/public/js/studio_edit.js:
```diff
@@ -0,0 +1,139 @@
+        cancel: function() {
+            // Remove TinyMCE instances to make sure jQuery does not try to access stale instances
+            // when loading editor for another block:
+            for (var i in fields) {
```
Why not use [`_.each()`](http://underscorejs.org/#each) or [`$.each()`](http://api.jquery.com/jQuery.each/) instead?

dseaton commented:
@smarnach I am fine with migrating String to Text. Just make sure to update any examples you are using as demoes.

smarnach commented:
@dseaton Your comment was meant for open-craft/xblock-activetable#1.  :)  I implemented the change and updated all examples.

e-kolpakov commented:
@singingwolfboy answering your question about numpy - this is mostly a conversion of https://github.com/open-craft/jsinput-vectordraw to an XBlock, so in order to fit this into the deadline the more is taken intact from it the better. We have discussed using numpy when implementing jsinput-vectordraw and the reason we haven't used

e-kolpakov commented:
Sorry, wrong PR :)

itsjeyd commented:
@singingwolfboy Thanks for reviewing! I addressed your comments. Additional changes are here: 0d4a06ac20e80f1e6ad9d0bb490c374d5cafcf39.

nedbat commented on .travis.yml:
```diff
@@ -0,0 +1,19 @@
+  - pip install -r test-requirements.txt
+  - pip install -r $VIRTUAL_ENV/src/xblock-sdk/requirements/base.txt
+  - pip install -r $VIRTUAL_ENV/src/xblock-sdk/requirements/test.txt
+  - pip install -r $VIRTUAL_ENV/src/xblock/requirements.txt
```
This seems fragile, because the xblock and xblock-sdk repos don't expect their requirements to be installed like this.  I would really like to understand why it's necessary to install these, and what the right way to install them is.

smarnach replied:
@nedbat Requirements files are flat.  The transitive requirements of packages listed in `requirements.txt` are not installed implicitly, unless they are specified in `install_requires` in `setup.py`.  However, only requirements available on PyPI can be implicitly installed in this way.  Since many of our requirements are installed from Github, we can only list them in `requirements.txt`, and we then have to resort to hacks like this one to install the transitive dependencies.

An alternative would be to flatten the list of recursive requirements manually, and list all of them in `requirements.txt` in this XBlock.  However, this is fragile as well, since then we have to update requirements in even more places in sync.

Uploading all versions of our packages to PyPI comes with its own set of problems, too.  I'm sure we could manage requirements in a better way than we currently do, but it isn't an easy problem.

bradenmacdonald replied:
@nedbat Are you saying we should use `cd $VIRTUAL_ENV/src/xblock-sdk/ && make pip` instead of the commands that we are using now (which are the same as what's in the Makefile)? Or something else?

nedbat replied:
@smarnach @bradenmacdonald I don't have a specific proposal.   I wanted to know why you needed to install xblock-sdk's test requirements.  I would have thought they were only needed for testing the sdk, which you are not doing here.  I understand about the difficulty of managing packages, and I'm sure part of this is us being sloppy in the xblock-sdk repo.  I wanted to understand the problem so that we could decide on a way to fix it, wherever the best place is.

bradenmacdonald replied:
@nedbat Our integration tests on Travis CI use bok choy and selenium to run through the various XBlock views/handlers hosted within the XBlock SDK "workbench" runtime. That means that we need to install XBlock, XBlock SDK, bok choy, mock, ddt, etc. But if we specify those requirements in our own XBlock's requirements.txt (whether using pinned versions or not), we'll inevitably arrive at a situation where our XBlock requires a different version of some dependency than xblock-sdk does, which leads to inconsistent test results depending on environment and order of installation. And for people like me who have an "xblock venv" where I install xblock-sdk plus about 20 different XBlocks that I work on, version mismatches can be frustrating time sinks.

So we've found that the easiest solution for now is just to say "we'll have what they're having" with respect to xblock-sdk, and standardize on whatever requirements it is using.

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+                "x1": int(float(line.get_attribute("x1"))),
+                "y1": int(float(line.get_attribute("y1"))),
+                "x2": int(float(line.get_attribute("x2"))),
+                "y2": int(float(line.get_attribute("y2"))),
```
Why are you first converting to `float`, and then converting to `int`? Seems like wasted computer cycles to me. If you want to strip off everything after the decimal point, you can do this instead:
```python
int(line.get_attribute("x1").split(".", 1)[0])
```

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+        expected_position = position.items()
+        for point in point_elements:
+            point_position = {
+                "cx": int(round(float(point.get_attribute("cx")))),
```
Same comment here. If you want to just get an `int` and discard everything after the decimal point, you can do this:
```python
int(line.get_attribute("x1").split(".", 1)[0])
```

singingwolfboy commented on tests/integration/test_vectordraw.py:
```diff
@@ -0,0 +1,1032 @@
+        self.assertTrue(status.is_displayed())
+        correctness = status.find_element_by_css_selector(".correctness")
+        if answer_correct:
+            self.assertTrue("checkmark-correct fa fa-check" in correctness.get_attribute("class"))
```
Use [`self.assertIn()`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertIn).

singingwolfboy commented on vectordraw/public/js/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+            selector = selector.split('-');
+            return {
+                type: selector[0],
+                idx: parseInt(selector[1])
```
Always pass the radix parameter to `parseInt()`.
```js
idx: parseInt(selector[1], 10)
```

singingwolfboy commented on vectordraw/public/js/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+        var length = vec_settings.length_factor * Math.sqrt(Math.pow(x2-x1, 2) + Math.pow(y2-y1, 2));
+        var angle = ((Math.atan2(y2-y1, x2-x1)/Math.PI*180) - vec_settings.base_angle) % 360;
+        if (angle < 0) {
+            angle += 360;
```
Could it ever be the case that the `angle` value is actually less than -360? I don't fully understand everything this code is doing, so maybe it's not possible. But if it could ever happen, then this `if` check should be a `while` check instead.

smarnach replied:
Chiming in since I just happened to see this comment.

No, this can't ever be less than -360, since the previous line reduces modulo 360, so the result is in the range -359 to 359.  (If JavaScript's modulo operator were sane, the result would always be in the range 0 to 359.  However, most languages get this wrong, with Python being an exception.)

singingwolfboy commented on vectordraw/public/js/vectordraw.js:
```diff
@@ -0,0 +1,633 @@
+
+    VectorDraw.prototype.getState = function() {
+        var vectors = {}, points = {};
+        this.settings.vectors.forEach(function(vec) {
```
You're calling the `.forEach()` method directly, which is not fully supported in Internet Explorer 9 and below. ([see caniuse](http://caniuse.com/#search=foreach)) However, Open edX only [supports Internet Explorer 10 and above](http://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/named-release-cypress/front_matter/browsers.html), so that's fine.

My question is, why do you call `.forEach()` directly here, and use `_.forEach()` in other places? The version in the Underscore library is just a proxy to the native version when the native version is defined, so there shouldn't be any performance difference. It's not a big deal, but the lack of consistency is a bit confusing, and I wanted to see if there was a reason that I don't know about.

itsjeyd replied:
@singingwolfboy No reason, this is really just an inconsistency. Fixed in 645115fbc267ce6485f4d6544bf5b4710523b648 (which also addresses the remaining comments you posted today).

singingwolfboy commented on vectordraw/public/js/vectordraw_edit.js:
```diff
@@ -0,0 +1,720 @@
+        var lineID = lineElement.attr("id");
+
+        var titleID = lineID + "-title";
+        var titleElement = $("<title>").attr("id", titleID).text(vec.name);
```
Nitpick: you can accomplish the same thing by doing:
```js
var titleElement = $("<title>", {"id": titleID, "text": vec.name});
```

singingwolfboy commented on vectordraw/templates/html/vectordraw.html:
```diff
@@ -0,0 +1,130 @@
+          <span class="reset-label" aria-hidden="true">{% trans "Reset" %}</span>
+          <span class="sr">{% trans "Reset board to initial state" %}</span>
+        </button>
+        <button class="redo" title="Redo">
```
The `title` attribute should be translated.

singingwolfboy commented on vectordraw/templates/html/vectordraw.html:
```diff
@@ -0,0 +1,130 @@
+          <span class="redo-label fa fa-repeat" aria-hidden="true"></span>
+          <span class="sr">{% trans "Redo last action" %}</span>
+        </button>
+        <button class="undo" title="Undo">
```
The `title` attribute should be translated.

singingwolfboy commented on vectordraw/templates/html/vectordraw_edit.html:
```diff
@@ -0,0 +1,372 @@
+                      id="xb-field-edit-{{field.name}}"
+                      aria-describedby="{{field.name}}-help">
+                <option value="1" {% if field.value %}selected{% endif %}>
+                  True {% if field.default %}&nbsp;&nbsp;&nbsp;&nbsp;(Default){% endif %}
```
"True" and "Default" should be translated.

singingwolfboy commented on vectordraw/templates/html/vectordraw_edit.html:
```diff
@@ -0,0 +1,372 @@
+                  True {% if field.default %}&nbsp;&nbsp;&nbsp;&nbsp;(Default){% endif %}
+                </option>
+                <option value="0" {% if not field.value %}selected{% endif %}>
+                  False {% if not field.default %}&nbsp;&nbsp;&nbsp;&nbsp;(Default){% endif %}
```
"False" and "Default" should be translated.

singingwolfboy commented on vectordraw/templates/html/vectordraw_edit.html:
```diff
@@ -0,0 +1,372 @@
+            WYSIWYG Editor
+            <button class="info-button" title="Info">
+              <span class="info-label fa fa-question" aria-hidden="true"></span>
+              <span class="sr">{% trans "Toggle information about how to use WYSIWYG editor" %}</span>
```
I think you're missing the word "the" -- "how to use *the* WYSIWYG editor". You also may want to add a translators comment explaining what WYSIWYG means.

singingwolfboy commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,544 @@
+        # Check vectors
+        vectors = data.get('vectors')
+        if not isinstance(vectors, dict):
+            return False
```
Might be better to have this function raise a `ValueError` when it finds an issue, rather than returning False.

itsjeyd commented:
@singingwolfboy I addressed all of your comments. Please let me know if there is anything else you would like me to address.

singingwolfboy commented on vectordraw/vectordraw.py:
```diff
@@ -0,0 +1,546 @@
+                raise ValueError
+        # If we get to this point, it means that vector and point data is valid;
+        # the only thing left to check is whether data contains a list of checks:
+        return 'checks' in data
```
Shouldn't this be:
```
if 'checks' not in data:
    raise ValueError
return True
```
Or do you mean to have this return either True or False?

itsjeyd replied:
@singingwolfboy Ah, yes, now that this method raises `ValueError` if it comes across invalid data, it doesn't actually need to return `True` or `False`. So this:

```python
if 'checks' not in data:
    raise ValueError
```
should be enough. Thanks!

singingwolfboy commented:
I only found one last little issue, and everything else looks good. After you address it, :+1:  from me. (I'll check the checkbox in the PR description, as well.)

itsjeyd commented:
@singingwolfboy That's great, thank you! The fix for the last issue is here: c03551d978a50be0e5ea91b510f1c21f7fed3ac3

cptvitamin commented on vectordraw/templates/html/vectordraw.html:
```diff
@@ -0,0 +1,130 @@
+  <div id="vectordraw">
+    <div class="menu" style="width: {{ self.width }}px;">
+      <div class="controls">
+        <span class="sr" id="element-list-add-label">{% trans "Select element to add to board" %}</span>
```
@itsjeyd can we make this a label element and change the aria-labelledby attribute below to a for attribute?  It's better supported and won't trigger test failures.

cptvitamin commented:
@itsjeyd can you make the :focus state on all of the buttons match the :hover state?  Currently, there is no way for a keyboard user to know what button the are currently focused on.

cptvitamin commented on vectordraw/public/css/vectordraw.css:
```diff
@@ -0,0 +1,172 @@
+
+/* Make sure screen-reader content is hidden in the workbench: */
+.vectordraw_block .sr {
+    display: none;
```
@itsjeyd you cannot have display:none on SR text.  display none is the same as aria-hidden, so this text is not exposed to a screen reader user.  all of the other styles below are effective for exposing it only to screen reader users.

cptvitamin commented:
@itsjeyd this is looking great.  I love the addition of the properties panel for keyboard only users.  I also think it will help users with dexterity issues as well as all users who want to position things precisely.  My only suggestion is that some sort of grid can be toggled into view.  While attempting to use the manual controls I found it very difficult to position vectors using trial and error.

nedbat commented:
:+1:

itsjeyd commented:
@cptvitamin Thanks for the feedback! I addressed your comments: 461371b21bff52337353792164c0ea8d7a84f58a. As for your suggestion about making available a grid to help keyboard users place vectors, this is already available as an opt-in feature. Course authors can set an "Axis" option to "True" in Studio to get something like this:

![axis-enabled](https://cloud.githubusercontent.com/assets/961441/11714281/57c1dd44-9f39-11e5-99e2-098ee45ffc0b.png)

To promote the use of this feature I changed the default for the "Axis" option to "True" (it was disabled by default before) and added a comment to the help text that points out that enabling the option makes an exercise more accessible for keyboard users. It would also be possible to add functionality for toggling the axis on demand but this would require more substantial changes (JSXGraph does not provide an API for toggling the axis after the drawing board has been created), so this is not something we can implement right now (considering that we are close to the merge deadline for this work). Please let me know if there is anything else I should address.

sarina commented:
Mark is out sick today, and possibly tomorrow as well, so I don't think we
will be able to get more from him until next week.

On Thu, Dec 10, 2015 at 6:31 AM, Tim Krones <notifications@github.com>
wrote:

> @cptvitamin <https://github.com/cptvitamin> Thanks for the feedback! I
> addressed your comments: 461371b
> <https://github.com/open-craft/xblock-vectordraw/commit/461371b21bff52337353792164c0ea8d7a84f58a>.
> As for your suggestion about making available a grid to help keyboard users
> place vectors, this is already available as an opt-in feature. Course
> authors can set an "Axis" option to "True" in Studio to get something like
> this:
>
> [image: axis-enabled]
> <https://cloud.githubusercontent.com/assets/961441/11714281/57c1dd44-9f39-11e5-99e2-098ee45ffc0b.png>
>
> To promote the use of this feature I changed the default for the "Axis"
> option to "True" (it was disabled by default before) and added a comment to
> the help text that points out that enabling the option makes an exercise
> more accessible for keyboard users. It would also be possible to add
> functionality for toggling the axis on demand but this would require more
> substantial changes (JSXGraph does not provide an API for toggling the axis
> after the drawing board has been created), so this is not something we can
> implement right now (considering that we are close to the merge deadline
> for this work). Please let me know if there is anything else I should
> address.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-craft/xblock-vectordraw/pull/1#issuecomment-163584783>
> .
>

smarnach commented:
@sarina Oh, that's really unfortunate!  We'd really like to get these XBlocks into the release, since the client wants to start using them, and the next release will be in January.  Is there anyone who could fill in for Mark?

Alternatively, since we already have done a full accessibility review cycle on these XBlocks, addressed Mark's comments and have +1s from the other code reviews, it might be reasonable to just merge them now given the circumstances.  If it turns out we need further accessibility tweaks, we could add them later, but things should be looking pretty good already.  On the ActiveTable XBlock, there are no open questions left and only a formal +1 is missing, and on this one it's only Tim's last comment that's waiting for an answer from Mark.

sarina commented:
OK, I would be fine with you merging this in, as long as you make certain
to follow up with Mark next week and address any final issues he may have.

On Thu, Dec 10, 2015 at 10:08 AM, Sven Marnach <notifications@github.com>
wrote:

> @sarina <https://github.com/sarina> Oh, that's really unfortunate! We'd
> really like to get these XBlocks into the release, since the client wants
> to start using them, and the next release will be in January. Is there
> anyone who could fill in for Mark?
>
> Alternatively, since we already have done a full accessibility review
> cycle on these XBlocks, addressed Mark's comments and have +1s from the
> other code reviews, it might be reasonable to just merge them now given the
> circumstances. If it turns out we need further accessibility tweaks, we
> could add them later, but things should be looking pretty good already. On
> the ActiveTable XBlock, there are no open questions left and only a formal
> +1 is missing, and on this one it's only Tim's last comment that's waiting
> for an answer from Mark.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-craft/xblock-vectordraw/pull/1#issuecomment-163653030>
> .
>

smarnach commented:
@sarina Thanks a lot!  That's a great relief for us. :)

itsjeyd commented:
@sarina Yes, and we'll make sure to follow up with Mark next week.
